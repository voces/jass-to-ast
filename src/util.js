import { SingleProp } from "./grammar/types.js";

const primitives = [String, Boolean];

const primitiveNode = (node) =>
	typeof node !== "object" ||
	node === null ||
	primitives.some((primitive) => node instanceof primitive);

export const inspect = (ast, enumerable = true, char = "\t", depth = 0) => {
	if (ast === undefined) return char.repeat(depth) + "__undefined__";

	if (ast === null) return char.repeat(depth) + "Null";

	if (primitiveNode(ast))
		return (
			char.repeat(depth) +
			ast.constructor.name +
			" " +
			JSON.stringify(ast)
		);

	if (ast instanceof SingleProp)
		return (
			char.repeat(depth) +
			ast.constructor.name +
			" " +
			inspect(ast.data[0], enumerable, char, depth).trim()
		);
	// const superClass = Object.getPrototypeOf( Object.getPrototypeOf( ast ) ).constructor;
	// if ( superClass === Node ) return ast.constructor.name + " " + inspect( ast.data, enumerable, char, depth, true );
	// return ast.constructor.name + " " + inspect( new superClass( ...ast.data ), enumerable, char, depth, true );

	if (ast instanceof Array) {
		let s = char.repeat(depth) + ast.constructor.name + " [\n";
		for (let i = 0; i < ast.length; i++)
			s +=
				char.repeat(depth + 1) +
				inspect(ast[i], enumerable, char, depth + 1).trim() +
				"\n";
		return s + char.repeat(depth) + "]";
	}

	const s = char.repeat(depth) + ast.constructor.name + " {";
	let body = "";
	if (enumerable)
		for (const prop in ast)
			body +=
				char.repeat(depth + 1) +
				prop +
				": " +
				inspect(ast[prop], enumerable, char, depth + 1).trim() +
				"\n";
	else {
		const props = Object.getOwnPropertyDescriptors(ast);
		for (const prop in props)
			body +=
				char.repeat(depth + 1) +
				prop +
				": " +
				inspect(ast[prop], enumerable, char, depth + 1).trim() +
				"\n";
	}
	return s + (body ? "\n" + body + char.repeat(depth) + "}" : "}");
};
