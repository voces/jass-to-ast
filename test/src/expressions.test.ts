import { parseSnapshot } from "../util";

it(
	"not without spaces",
	parseSnapshot(`
		function f takes nothing returns boolean
		if(not(udg_a==udg_b))then
			return false
		endif
		endfunction
	`),
);
