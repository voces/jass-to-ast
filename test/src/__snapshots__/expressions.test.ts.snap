// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`not without spaces 1`] = `
"Program [
	JASSFunction {
		name: String \\"f\\"
		returns: String \\"boolean\\"
		statements: Statements [
			IfThenElse {
				condition: Parens UnaryOp {
					operator: String \\"not\\"
					expr: Parens BinaryOp {
						left: Name \\"udg_a\\"
						operator: String \\"==\\"
						right: Name \\"udg_b\\"
					}
				}
				then: Statements [
					Return Boolean false
				]
			}
		]
	}
]"
`;
